zipline/pipeline/loaders/utils.py

```diff
@@ -272,3 +274,116 @@ def check_data_query_args(data_query_time, data_query_tz):
                 data_query_tz,
             ),
         )
+
+
+def last_in_date_group(df, reindex, dates, assets, have_sids=True,
+                       extra_groupers=[]):
+    """
+    Determine the last piece of information known on each date in the date
+    index for each group.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        The DataFrame containing the data to be grouped.
+    reindex : bool
+        Whether or not the DataFrame should be reindexed against the date
+        index. This will add back any dates to the index that were grouped
+        away.
+    dates : pd.DatetimeIndex
+        The dates to use for grouping and reindexing.
+    assets : pd.Int64Index
+        The assets that should be included in the column multiindex.
+    have_sids : bool
+        Whether or not the DataFrame has sids. If it does, they will be used
+        in the groupby.
+    extra_groupers : list of str
+        Any extra field names that should be included in the groupby.
+
+    Returns
+    -------
+    last_in_group : pd.DataFrame
+        A DataFrame with dates as the index and fields used in the groupby as
+        levels of a multiindex of columns.
+
+    """
+    idx = [dates[dates.searchsorted(
+        df[TS_FIELD_NAME].values.astype('datetime64[D]')
+    )]]
+    if have_sids:
+        idx += [SID_FIELD_NAME]
+    idx += extra_groupers
+
+    last_in_group = df.drop(TS_FIELD_NAME, axis=1).groupby(
+        idx,
+        sort=False,
+    ).last()
+
+    # For the number of things that we're grouping by (except TS), unstack
+    # the df
+    last_in_group = last_in_group.unstack(list(range(-1, -len(idx), -1)))
+
+    if reindex:
+        if have_sids:
+            cols = last_in_group.columns
+            last_in_group = last_in_group.reindex(
+                index=dates,
+                columns=pd.MultiIndex.from_product(
+                    tuple(cols.levels[0:len(extra_groupers) + 1]) + (assets,),
+                    names=cols.names,
+                ),
+            )
+        else:
+            last_in_group = last_in_group.reindex(dates)
+
+    return last_in_group
+
+
+def ffill_across_cols(df, columns, name_map):
```

Comment:
Not in all cases. This would work for the events loader, but for the estimates loader, the datasets of the columns passed in are built dynamically, and so we need to look at the parent to get the right name mapping.